### DISCUSSION TITLE
Add API for management of access to security alerts

### DISCUSSION THREAD

**Discussion by reedloden** [community/community#14021] [Code Security]
Requesting that an API be added for management of access to security alerts

See https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-security-and-analysis-settings-for-your-repository#granting-access-to-security-alerts

Right now, users/teams have to be manually added via that UI, and with dozens if not hundreds of repositories under an organization/enterprise, that is a real maintenance burden. Would be nice to be able to use an API to automate it.

Until there's an API available, the closest I have been able to find is a script somebody wrote that just abuses the standard web interface for this purpose: https://gist.github.com/captn3m0/8806a2b7943657c39cc42502560a0f9f.

**Comment by greysteil** [comment]
Hey @reedloden, thanks for the feedback.

We're planning to remove the "access to alerts" UI in the not-too-distant future (once its functionality has been fully replaced with fine grained permissions and custom roles), so we aren't planning to add an API for it. Unfortunately for now that puts us in an in-between place, which I appreciate is frustrating.

In addition to the script above, if you're a GitHub Enterprise customer you can use [custom repository roles](https://docs.github.com/en/enterprise-cloud@latest/organizations/managing-peoples-access-to-your-organization-with-roles/managing-custom-repository-roles-for-an-organization) to achieve a similar result:
1. [Create a custom repository role](https://docs.github.com/en/enterprise-cloud@latest/organizations/managing-peoples-access-to-your-organization-with-roles/managing-custom-repository-roles-for-an-organization) inheriting from `write` and add all of the permissions labelled with `security` to it. Call it "Write with security" or something like that
2. For each repo / user pair that you want to grant access to, use the [add a repository collaborator](https://docs.github.com/en/rest/reference/collaborators#add-a-repository-collaborator) API endpoint to update that user's permission level on the repository to be the new custom role
3. Alternatively you can add many users to a team ([via the API](https://docs.github.com/en/rest/reference/teams#add-or-update-team-membership-for-a-user)) and then use the team [API to add or update that team's permission](https://docs.github.com/en/rest/reference/teams#add-or-update-team-repository-permissions) level on the repository to be the new custom role

A couple of caveats:
* The above is (currently) only possible for GitHub Enterprise customers (on server you need to be on GHES 3.5+)
* Custom roles are only available at the organisation level. If you have many organisations you'll need to manually create the same role on each organisation via the UI, which I appreciate is no fun

We're keeping the "access to alerts" UI for now, because it works slightly differently to the above (it's composable with a user's role on a repo and is available to customers without an Enterprise account) but for Enterprise customers, ☝️ is the way forward.

Finally, if you who want members of a team to have access to security alerts on **all repositories** in an organisation, the [security manager role](https://github.blog/changelog/2021-10-21-introducing-the-organization-level-security-manager-role/) is the way to go.

Apologies for the imperfect solution, and particularly that it's not available for all customers. I'm also cc-ing @hpsin who leads our wider work on identity, and will be the one who fully solves this in the future! :octocat:

**Comment by glye** [comment]
The same goes for granting access to security advisories. I'd like to always give certain users or groups access to draft advisories for given repos.

**Reply by greysteil** [reply]
Makes sense, and that's exactly why we want to solve this with fine grained permissions and the wider GitHub identity and access management functionality.

Cc @KateCatlin @jhutchings1 - it looks like we don't currently have fine grained permissions for managing security advisories. Can you take a look?

**Reply by KateCatlin** [reply]
100%! This is a common issue that we're hearing about in user interviews, and we're looking into changing these permissions in the next few quarters. 

@glye @reedloden would love to get you on the phone and chat through your user stories/use cases here. Would you be open to a 30-minute zoom? Happy to send you a gift card as a thank you: https://calendly.com/security-advisories-ux-calls/30min?month=2022-11

**Reply by glye** [reply]
@KateCatlin Sure! See you next week, then. Thanks for dealing with this.

**Comment by ranma2913** [comment]
Special Thanks for this. 

I was able to create a new role from Triage and it gave the correct access to my group's Managers & Systems-Analyst. By choosing Triage we were able to maintain our separation of duties compliance keeping those with production access away from having Write permission to the codebase.

I chose:

- Role Name: `triage-security-alerts`
- Description: `Triage access plus permission to view and manage code scanning, Dependabot, or secret scanning alerts.`
- Inherit `Triage`
- Add all 'security' permissions 

![image](https://github.com/community/community/assets/4295880/f59c2b18-6cc4-4850-97ce-33feb6573a01)

**Comment by beastmode846** [comment]
{pageName:SignIn,jsPath:[],pageProps:{signInForm:{values:{email:",password:",rememberEmail:false},errors:{},csrfToken:8d2409a9eca0551c940b6299f258eb7bf9d4b360-1691608105182-cf4b264158c71ca1eff22cc2},redirect:ff850b813f59dd4ca8d95c0a381ad43378a9634a-1691608105005-/mobile/history/G4AErN/tradeVolume.xlsx?header=false,sandboxMode:false,oauthClient:null,siteKeyV2:6LdMgEwgAAAAAOdD5qV6rxFJ9pYSgO0GPUAyv_jO,siteKeyV3:6LeXFEEgAAAAALds-drHZpoMktvuxcTY1NE1ChO2},templateProps:{isSandbox:false,featureFlags:{GeminiCreditCardEmailVerificationEnabled:true,EuropeAgreementsEnabled:true,TradePerpEnabled:true,PerpsUiEnabled:true,IrelandAgreementsEnabled:true,SardineDeviceIntelligenceEnabled:true,GeminiCreditCardEnabled:true}},sentry:{publicKey:abfac61979bc4c90aeb4b67a59ed0e61,projectId:7,version:dcbf78cce4fc,activeTrader:{publicKey:a4164a4062dc4fb5afae25221904fdd4,projectId:6104215,host:o472445},creditCard:{publicKey:dd02a86c535b4805b0c5cecaf2a13555,projectId:4504973743947776,host:o472445}},mixpanel:{token:d63e727b7647b63c44007c4c8876cb81},braze:{appIdentifierWeb:f7c7991a-0419-4e08-acae-40e64d232a89,sdkEndpoint:sdk.iad-03.braze.com},sift:{beaconToken:0490ca18f0},sardine:{clientId:87271e93-ae9b-4ed1-b98b-0ef903f22353},zendesk:{publicKey:79138343-14c5-436c-8fdf-67c1e0a6e392},datadog:{tradingDashBoardApplicationId:7ad4a764-8ab3-4513-a8f5-28287a2512ac,tradingDashBoardClientToken:pubf057b21f890b02dc00fc616bbaefb67c}}

**Reply by ranma2913** [reply]
This minified json is hard to read and has no additional context. What are you talking a bout?

**Comment by vini-ppro** [comment]
> We're planning to remove the "access to alerts" UI in the not-too-distant future

@greysteil - any news on that? It's almost 2025 :p

**Reply by greysteil** [reply]
Hey @vini-ppro, I'm afraid I'm no longer at GitHub, so I don't have any insight into this one. Hopefully one of the team will reply and let you know the plans for fine grained permissions / changes to the access to alerts UI.

**Comment by hussein-mimi** [comment]
@greysteil who can help in this case, i want to create repo using terraform, but it seems the only way to modify access alert is using github UI
